model/riscv_insts_vext_arith.sail

```diff
@@ -0,0 +1,554 @@
+                                vs2_val[idx]
+                              } else zeros()                   
+                            }
+      }
+    }
+  };
+
+  write_vreg(num_elem, SEW, LMUL_pow, vd, result);
+  vstart = EXTZ(0b0);
+  RETIRE_SUCCESS
+}
+
+mapping vvtype_mnemonic : vvfunct6 <-> string = {
+  VV_VADD          <-> "vadd.vv",
+  VV_VSUB          <-> "vsub.vv",
+  VV_VAND          <-> "vand.vv",
+  VV_VOR           <-> "vor.vv",
+  VV_VXOR          <-> "vxor.vv",
+  VV_VRGATHER      <-> "vrgather.vv",
+  VV_VRGATHEREI16  <-> "vrgatherei16.vv",
+  VV_VSADDU        <-> "vsaddu.vv",
+  VV_VSADD         <-> "vsadd.vv",
+  VV_VSSUBU        <-> "vssubu.vv",
+  VV_VSSUB         <-> "vssub.vv",
+  VV_VSLL          <-> "vsll.vv",
+  VV_VSMUL         <-> "vsmul.vv",
+  VV_VSRL          <-> "vsrl.vv",
+  VV_VSRA          <-> "vsra.vv",
+  VV_VSSRL         <-> "vssrl.vv",
+  VV_VSSRA         <-> "vssra.vv",
+  VV_VMINU         <-> "vminu.vv",
+  VV_VMIN          <-> "vmin.vv",
+  VV_VMAXU         <-> "vmaxu.vv",
+  VV_VMAX          <-> "vmax.vv"
+}
+
+mapping clause assembly = VVTYPE(funct6, vm, vs2, vs1, vd)
+  <-> vvtype_mnemonic(funct6) ^ spc() ^ vreg_name(vd) ^ sep() ^ vreg_name(vs2) ^ sep() ^ vreg_name(vs1) ^ maybe_vmask(vm)
+
+
+/* *******************************OPIVX(Vector Slide & Gather Instructions)******************************** */
+/* Slide and gather instructions extend rs1 / imm to XLEN bits (not SEW bits as others do) */
+
+union clause ast = VXSG : (vxsgfunct6, bits(1), regidx, regidx, regidx)
+
+mapping encdec_vxsgfunct6 : vxsgfunct6 <-> bits(6) = {
+  VX_VSLIDEUP     <-> 0b001110,
+  VX_VSLIDEDOWN   <-> 0b001111,
+  VX_VRGATHER     <-> 0b001100
+}
+
+mapping clause encdec = VXSG(funct6, vm, vs2, rs1, vd)
+  <-> encdec_vxsgfunct6(funct6) @ vm @ vs2 @ rs1 @ 0b100 @ vd @ 0b1010111
+
+function clause execute(VXSG(funct6, vm, vs2, rs1, vd)) = {
+  let SEW_pow  : int = get_sew_pow();
+  let SEW      : int = int_power(2, SEW_pow);
+  let LMUL_pow : int = get_lmul_pow();
+  let VLEN_pow : int = get_vlen_pow();
+  let num_elem : int = get_num_elem(LMUL_pow, SEW);
+
+  assert(8 <= SEW & SEW <= 64);
+  let 'n = num_elem;
+  let 'm = SEW;
+
+  let vm_val  : vector('n, dec, bool)     = read_vmask(num_elem, vm, vreg_name("v0"));
+  let rs1_val : int                       = unsigned(X(rs1));
+  let vs2_val : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vs2);
+  let vd_val  : vector('n, dec, bits('m)) = read_vreg(num_elem, SEW, LMUL_pow, vd);
+  result      : vector('n, dec, bits('m)) = undefined;
+  mask_helper : vector('n, dec, bool)     = undefined;
+
+  (result, mask_helper) = init_masked_result(num_elem, SEW, LMUL_pow, vd_val, vm_val);
+
+  foreach (i from 0 to (num_elem - 1)) {
+    if mask_helper[i] == true then {
+      result[i] = match funct6 {
+        VX_VSLIDEUP    => {
+                            assert(vs2 != vd);
+                            assert(i - rs1_val < 'n);
```

Comment:
Not sure this is correct as all possible rs1_values are valid. A larger or smaller rs1_val only affects which element are kept unchanged in vd.

Reply:
This assert is also required by the compiler for this operation `vs2_val[i - rs1_val]` in the next line. In fact, it will be true for all `rs1_val` values because `i` is always smaller than `'n` and `rs1_val` is an unsigned integer.

Reply:
I don't think you need the `: int` type annotation on rs1_val. The type checker will infer precise bounds from the bit width. The int annotation will just override that.

Reply:
This assert statement has been removed.